dhall-docs/src/Dhall/Docs/Comment.hs

```diff
@@ -0,0 +1,305 @@
+{-| Provides utilities to parse Dhall comments
+-}
+
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ViewPatterns      #-}
+
+module Dhall.Docs.Comment
+    ( parseComments
+    , CommentParseError(..)
+    , DhallDocsText
+    , parseSingleDhallDocsComment
+    , unDhallDocsText
+    ) where
+
+import Control.Applicative (many, some, (<|>))
+import Data.Functor        (void)
+import Data.List.NonEmpty  (NonEmpty (..), (<|))
+import Data.Text           (Text)
+import Dhall.Docs.Util
+import Dhall.Parser        (Parser (..))
+import Text.Megaparsec     (SourcePos, (<?>))
+
+import qualified Data.Either
+import qualified Data.Foldable
+import qualified Data.List.NonEmpty       as NonEmpty
+import qualified Data.Maybe               as Maybe
+import qualified Data.Text
+import qualified Dhall.Parser.Combinators as Combinators
+import qualified Dhall.Parser.Expression  as Expression
+import qualified Dhall.Parser.Token       as Token
+import qualified Text.Megaparsec
+import qualified Text.Megaparsec.Pos      as Megaparsec.Pos
+
+-- | For explanation of these data-types see 'DhallComment'
+data DhallDocsComment
+data MarkedComment
+data RawComment
+
+type ListOfSingleLineComments = NonEmpty (SourcePos, Text)
+
+{-| Internal representation of Dhall comments. Text is always stripped from
+    whitespace from both the start and begin of the string
+
+    * If @a = 'DhallDocsComment'@ then comment is valid to extract its contents
+    and be rendered on documentation
+    * If @a = 'MarkedComment'@ then comment has the @|@ marker that
+    @dhall-docs@ will be aware of, but this comment may not be a @dhall-docs@
+    comment
+    * If @a = `RawComment`@ then the comment is a raw comment
+-}
+data DhallComment a
```

Comment:
If you enable the `DataKinds` extension, you can do this:

```haskell
data CommentType = DhallDocsComment | MarkedComment | RawComment

data DhallComment (a :: CommentType)
```

... which ensures that `a` can only be one of those three constructors

Reply:
I had to also add `KindSignatures` to properly use the type restriction on `DhallComment